qiskit/algorithms/minimum_eigen_solvers/qaoa/qaoa_circuit.py

```diff
@@ -0,0 +1,209 @@
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Global X phases and parameterized problem hamiltonian."""
+
+from typing import Optional, Union, cast, Set
+
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.circuit import Parameter, ParameterVector
+from qiskit.circuit.library import BlueprintCircuit
+from qiskit.opflow import OperatorBase, I, X, CircuitStateFn, H, CircuitOp, EvolutionFactory
+
+
+class QAOACircuit(BlueprintCircuit):
+    """Global X phases and parameterized problem hamiltonian."""
+
+    def __init__(self,
+                 cost_operator: OperatorBase,
+                 param_p: int,
+                 initial_state: Optional[QuantumCircuit] = None,
+                 mixer_operator: Optional[Union[QuantumCircuit, OperatorBase]] = None,
+                 name: str = "qaoa"):
+        """
+        Constructor, following the QAOA paper https://arxiv.org/abs/1411.4028
+
+        Args:
+            cost_operator: The operator representing the cost of the optimization problem,
+                denoted as U(B, gamma) in the original paper.
+            param_p: The integer parameter p, which determines the depth of the circuit,
+                as specified in the original paper.
+            initial_state: An optional initial state to use.
+            mixer_operator: An optional custom mixer to use instead of the global X-rotations,
+                denoted as U(B, beta) in the original paper. Can be an operator or
+                an optionally parameterized quantum circuit.
+            name: A name of the circuit, default 'qaoa'
+        Raises:
+            TypeError: invalid input
+        """
+        super().__init__(name=name)
+        self._cost_operator = cost_operator
+        self._param_p = param_p
+        self._initial_state = initial_state
+        self._mixer_operator = mixer_operator
+
+        # set this circuit as a not-built circuit
+        self._num_qubits = 0
+        self._num_parameters = 0
+        self._bounds = None
+        self._mixer = None
+        self._data = None
+
+    def _check_configuration(self, raise_on_failure: bool = True) -> bool:
+        valid = True
+
+        if self._cost_operator is None:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The operator representing the cost of "
+                                     "the optimization problem is not set")
+        if self._param_p is None or self._param_p < 1:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The integer parameter p, which determines the depth "
+                                     "of the circuit, either not set or set to non-positive value")
```

Comment:
Similar comment to cost_operator in terms of it being None and doing check of >= 1 in setter

Reply:
Same as above.